libraries/custom_appbase/include/eosio/chain/exec_pri_queue.hpp

```diff
@@ -61,22 +76,56 @@ class exec_pri_queue : public boost::asio::execution_context
       should_exit_ = [](){ assert(false); return true; }; // should not be called when locking is disabled
    }
 
-   // called from appbase::application_base::exec poll_one() or run_one()
    template <typename Function>
-   void add(int priority, exec_queue q, size_t order, Function function) {
+   void add(int priority, exec_queue q, size_t order, Function&& function) {
       assert( num_read_threads_ > 0 || q != exec_queue::read_exclusive);
       prio_queue& que = priority_que(q);
-      std::unique_ptr<queued_handler_base> handler(new queued_handler<Function>(priority, order, std::move(function)));
+      std::unique_ptr<queued_handler_base> handler(new queued_handler<Function>(handler_id::unique, priority, order, std::forward<Function>(function)));
       if (lock_enabled_ || q == exec_queue::read_exclusive) { // called directly from any thread for read_exclusive
          std::lock_guard g( mtx_ );
-         que.push( std::move( handler ) );
+         que.push( handler.release() );
          if (num_waiting_)
             cond_.notify_one();
       } else {
-         que.push( std::move( handler ) );
+         que.push( handler.release() );
       }
    }
 
+   // called from appbase::application_base::exec poll_one() or run_one()
+   template <typename Function>
+   void add(handler_id id, int priority, exec_queue q, size_t order, Function&& function) {
+      assert( num_read_threads_ > 0 || q != exec_queue::read_exclusive);
+      if (id == handler_id::unique) {
+         return add(priority, q, order, std::forward<Function>(function));
+      }
+      prio_queue& que = priority_que(q);
+      std::unique_lock g( mtx_, std::defer_lock );
+      if (lock_enabled_ || q == exec_queue::read_exclusive) {
+         // called directly from any thread for read_exclusive
+         g.lock();
+      }
+      if (!que.empty()) {
+         // find the associated priority
+         auto end = que.ordered_end();
+         auto i = std::lower_bound(que.ordered_begin(), end, priority, [](const auto& h, int priority) {
+            return h->priority() > priority;
+         });
+         if (i != end) {
+            // ordered iterator appears to only be a forward iterator
+            auto p = i;
+            for (; i != end; p = i, ++i) {
+               if ((*i)->priority() != priority)
+                  break;
+            }
```

Comment:
This seems wrong to me. Shouldn't we want to skip handlers which have the same priority, but a different id?

Reply:
The idea is the find the last posted function. If it same as you are attempting to post, don't bother.